pyro/infer/mcmc/hmc.py

```diff
@@ -343,38 +333,50 @@ def _initialize_model_properties(self):
                                                           self._has_enumerable_sites,
                                                           self.max_plate_nesting)
         mass_matrix_size = sum(self._r_numels.values())
-        if self.full_mass:
-            initial_mass_matrix = eye_like(site_value, mass_matrix_size)
-        else:
+        if self._adapter.is_diag_mass:
             initial_mass_matrix = site_value.new_ones(mass_matrix_size)
-        self._adapter.inverse_mass_matrix = initial_mass_matrix
+        else:
+            initial_mass_matrix = eye_like(site_value, mass_matrix_size)
+        self._adapter.configure(self._warmup_steps,
+                                inv_mass_matrix=initial_mass_matrix,
+                                find_reasonable_step_size_fn=self._find_reasonable_step_size)
+
+    def _initialize_sampling(self, trace):
+        z = {name: node["value"].detach() for name, node in self._iter_latent_nodes(trace)}
+        # automatically transform `z` to unconstrained space, if needed.
+        for name, transform in self.transforms.items():
+            z[name] = transform(z[name])
+        potential_energy = self._potential_energy(z)
+        self._cache(z, potential_energy, None)
+        if self._adapter.adapt_step_size:
+            self._adapter.reset_step_size_adaptation()
```

Comment:
Hmm..why don't we do any state setup that is trace dependent, when we set the initial trace instead? Would suggest a minor change:
 - Renaming this to `_initialize_step_size` to make the objective clearer.
 - Calling this from inside `initial_trace` instead of inside `sample`. Note that you will need to call this from both the setter and the getter.

Reply:
Nice solution! Thanks @neerajprad !